core/src/main/java/org/elasticsearch/indices/IndexingMemoryController.java

```diff
@@ -158,85 +168,213 @@ ByteSizeValue indexingBufferSize() {
-        }
+    /** returns how many bytes this shard is currently writing to disk */
+    protected long getShardWritingBytes(IndexShard shard) {
+        return shard.getWritingBytes();
+    }
+
+    /** ask this shard to refresh, in the background, to free up heap */
+    protected void writeIndexingBufferAsync(IndexShard shard) {
+        threadPool.executor(ThreadPool.Names.REFRESH).execute(new Runnable() {
+            @Override
+            public void run() {
+                shard.writeIndexingBuffer();
+            }
+        });
     }
 
-    /** check if any shards active status changed, now. */
+    /** force checker to run now */
     void forceCheck() {
         statusChecker.run();
     }
 
-    class ShardsIndicesStatusChecker implements Runnable {
+    /** called by IndexShard to record that this many bytes were written to translog */
+    public void bytesWritten(int bytes) {
+        statusChecker.bytesWritten(bytes);
+    }
+
+    /** Asks this shard to throttle indexing to one thread */
+    protected void activateThrottling(IndexShard shard) {
+        shard.activateThrottling();
+    }
+
+    /** Asks this shard to stop throttling indexing to one thread */
+    protected void deactivateThrottling(IndexShard shard) {
+        shard.deactivateThrottling();
+    }
+
+    @Override
+    public void postIndex(Engine.Index index) {
+        bytesWritten(index.getTranslogLocation().size);        
+    }
+
+    @Override
+    public void postDelete(Engine.Delete delete) {
+        bytesWritten(delete.getTranslogLocation().size);        
+    }
+
+    private static final class ShardAndBytesUsed implements Comparable<ShardAndBytesUsed> {
+        final long bytesUsed;
+        final IndexShard shard;
+
+        public ShardAndBytesUsed(long bytesUsed, IndexShard shard) {
+            this.bytesUsed = bytesUsed;
+            this.shard = shard;
+        }
+
         @Override
-        public synchronized void run() {
-            List<IndexShard> availableShards = availableShards();
-            List<IndexShard> activeShards = new ArrayList<>();
-            for (IndexShard shard : availableShards) {
-                if (!checkIdle(shard)) {
-                    activeShards.add(shard);
+        public int compareTo(ShardAndBytesUsed other) {
+            // Sort larger shards first:
+            return Long.compare(other.bytesUsed, bytesUsed);
+        }
+    }
+
+    /** not static because we need access to many fields/methods from our containing class (IMC): */
+    final class ShardsIndicesStatusChecker implements Runnable {
+
+        final AtomicLong bytesWrittenSinceCheck = new AtomicLong();
+        final ReentrantLock runLock = new ReentrantLock();
+
+        /** Shard calls this on each indexing/delete op */
+        public void bytesWritten(int bytes) {
+            long totalBytes = bytesWrittenSinceCheck.addAndGet(bytes);
+            if (totalBytes > indexingBuffer.bytes()/30) {
```

Comment:
I think we have to make this a `while (totalBytes > indexingBuffer.bytes()/30)` and then if we can't acquire the lock we step out. otherwise we might miss a required run?